### ISSUE TITLE
[Bug] Classification prediction varies according to the batch size

### ISSUE THREAD

**Issue by valentinecesbio** [cornellius-gp/gpytorch#2332]
# 🐛 Bug

I am using a multi-class GP model to perform classification (`gpytorch.models.ApproximateGP`) with the variational strategy `gpytorch.variational.LMCVariationalStrategy` and the variational distribution `gpytorch.variational.CholeskyVariationalDistribution`.

The main idea is to perform classification in order to produce a land cover map such as:

![image](https://user-images.githubusercontent.com/77112599/235616150-dcdb62aa-7eb1-445b-9127-8b9a23139354.png)

Each color corresponds to a label.

The classification predictions are changing by varying the batch size (`new_batch_size`)  from 16 to 2048. We also tried to changed the number of samples to draw from a latent GP when computing a likelihood (`new_number_samples`) from 10 to 100. A `torch.random.manual_seed(2)` was implemented in order to fully reproduce it.

| Batch size /     Num likelihood              | 10               |50                        | 100               |
|------------------------|------------------------------------|------------------------------------|------------------------------------|
| 16 | ![16-10](https://user-images.githubusercontent.com/77112599/235620224-6a58bf1d-92d5-4063-8d58-66f7f0d05553.png)|![16-50](https://user-images.githubusercontent.com/77112599/235620243-0ff15bf3-c53a-4306-bf7b-3fe24bc63b92.png)|![16-100](https://user-images.githubusercontent.com/77112599/235620286-b299f837-1b5d-48f0-96b0-2d707af8ea36.png)| 
| 128 |![128-10](https://user-images.githubusercontent.com/77112599/235620305-37126b8a-ecd6-472f-9651-2e0970ae88cc.png) | ![128-50](https://user-images.githubusercontent.com/77112599/235620317-e9f31b01-198f-409a-b112-0179daaf446b.png)|![128-100](https://user-images.githubusercontent.com/77112599/235620330-1d1f9561-b85c-4dbe-b02d-ef69d6d09548.png)| 
| 512 | ![512-10](https://user-images.githubusercontent.com/77112599/235620336-090d8b00-3c90-4cf8-82d4-165d73467d07.png) | ![512-50](https://user-images.githubusercontent.com/77112599/235620348-971a4808-6db1-4f4b-8fa0-ca99aacf3d0b.png)|![512-100](https://user-images.githubusercontent.com/77112599/235620363-bb1dc0fb-730b-4f35-833f-00c9ea15e0b7.png)| 
| 1024 |  ![1024-10](https://user-images.githubusercontent.com/77112599/235620439-a09e8b2c-3088-4dd1-a433-3ce71c93904e.png)| ![1024-50](https://user-images.githubusercontent.com/77112599/235620373-42c8b90c-7378-4077-8de7-c6322f33ceb1.png)|![1024-100](https://user-images.githubusercontent.com/77112599/235620542-7af9ed32-5866-498d-8ab0-aa62e44242b4.png)| 
| 2048 | ![2048-10](https://user-images.githubusercontent.com/77112599/235620610-60dbef85-1009-4948-a5e9-95740223afe8.png) | ![2048-50](https://user-images.githubusercontent.com/77112599/235620620-10ba2126-d705-4325-ba51-f79b774c11d3.png)|![2048-100](https://user-images.githubusercontent.com/77112599/235620637-8a23f14d-458f-48ee-831e-6789e6cf8195.png)| 
## To reproduce

**Data set**

Training dataset : `X_train_scaled` and `y_train_encode`

`X_train_scaled` is scaled using sklearn.preprocessing.StandardScaler

The scaler is saved in the file: `files/data.pt`

`y_train_encode` is encoded using sklearn.preprocessing.LabelEncoder

The encoder is saved in the file: `files/data.pt`

**Training**

Modified and simplified training script. Just to see how it was done. 

```python
def training(model, likelihood, train_loader, y_train_encode, learning_rate, number_epochs):

    optimizer = torch.optim.Adam([
        {'params': model.parameters()},
        {'params': likelihood.parameters()},
    ], lr=learning_rate)

   
    mll = gpytorch.mlls.VariationalELBO(likelihood, model,num_data=y_train_encode.size(0))

    model.cuda()
    model.train()
    likelihood.train()

    for num_epoch in range(number_epochs):

        for x_batch, y_batch in train_loader:
            optimizer.zero_grad()
            output = model(x_batch)
            loss = -mll(output, y_batch)
            loss.backward()
            optimizer.step()
```

```python
learning_rate = 0.01
number_epochs = 100
batch_size = 1024
learning_inducing_points = True
nb_classes = np.unique(y_train_encode).size   
nb_gp = nb_classes

# Extract random inducing points
torch.manual_seed(0)
perm = torch.randperm(X_train_scaled.size(0))
idx = perm[:number_inducing_points]
inducing_points = X_train_scaled[idx, :]
# Initialise each GP with the same inducing points
inducing_points = np.repeat(inducing_points[np.newaxis, :, :],nb_gp, axis=0)

model  = GP_MultiClass(inducing_points, learning_inducing_points, nb_classes, nb_gp)

likelihood = gpytorch.likelihoods.SoftmaxLikelihood(
                        num_features=
                        nb_gp,
                        num_classes=
                        nb_classes,
                        mixing_weights=
                        False)

dataset = torch.utils.data.TensorDataset(X_train_scaled, y_train_encode)
train_loader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, shuffle=True)

training(model, likelihood, train_loader, y_train_encode, learning_rate, number_epochs)

```

**Prediction function**

The model was trained before and saved in the file `"files/model.pt"`. The prediction is done using the saving model.

The following code is fully reproducible:

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import torch
import os
import gpytorch
import numpy as np
import rasterio
from rasterio.plot import reshape_as_raster, reshape_as_image
import json

class GP_MultiClass(gpytorch.models.ApproximateGP):

    def __init__(self, inducing_points, learning_inducing, nb_classes, nb_gp,
                 same_inducing_points, independant, init_lengthscale_temporal, SVDKL=False, learning_outputscale=True):

        # Define a Variational distribution
        variational_distribution = gpytorch.variational\
            .CholeskyVariationalDistribution(inducing_points.size(-2),
                                             batch_shape=torch.Size([nb_gp]))

        if independant:
            # MultitaskVariationalStrategy is deprecated, use
            # IndependentMultitaskVariationalStrategy instead.
            variational_strategy = gpytorch.variational.IndependentMultitaskVariationalStrategy(
                gpytorch.variational.VariationalStrategy(
                    self, inducing_points, variational_distribution,
                    learn_inducing_locations=learning_inducing
                ),
                num_tasks=nb_gp,
                )
        else:
            variational_strategy = gpytorch.variational.LMCVariationalStrategy(
                gpytorch.variational.VariationalStrategy(
                    self, inducing_points, variational_distribution,
                    learn_inducing_locations=learning_inducing
                ),
                num_tasks=nb_classes,
                num_latents=nb_gp,
                latent_dim=-1,
            )

        super(GP_MultiClass, self).__init__(variational_strategy)
        self.mean_module = gpytorch.means.ConstantMean(
            batch_shape=torch.Size([nb_gp]))

        if init_lengthscale_temporal == "ard":
            self.covar_module = gpytorch.kernels.ScaleKernel(
                gpytorch.kernels.RBFKernel(ard_num_dims=inducing_points.size(-1),
                                           batch_shape=torch.Size(
                                               [nb_gp])),
                batch_shape=torch.Size([nb_gp]), ard_num_dims=None)
        elif learning_outputscale:
            self.covar_module = gpytorch.kernels.ScaleKernel(
                gpytorch.kernels.RBFKernel(
                    batch_shape=torch.Size([nb_gp])),
                batch_shape=torch.Size([nb_gp]))

            if init_lengthscale_temporal is not None:

                if init_lengthscale_temporal == "mean":
                    if not same_inducing_points:
                        norme = torch.nn.functional.pdist(
                            inducing_points[0, :, :])
                    else:
                        norme = torch.nn.functional.pdist(
                            inducing_points[:, :])
                    mean = torch.mean(norme).item()
                    lengthscale = torch.as_tensor(mean)
                if init_lengthscale_temporal == "median":
                    if not same_inducing_points:
                        norme = torch.nn.functional.pdist(
                            inducing_points[0, :, :])
                    else:
                        norme = torch.nn.functional.pdist(
                            inducing_points[:, :])
                    median = torch.median(norme).item()
                    lengthscale = torch.as_tensor(median)
                if init_lengthscale_temporal == "dim":
                    dim = np.sqrt(inducing_points.size(-1))
                    lengthscale = torch.as_tensor(dim)

                self.covar_module.base_kernel.lengthscale = lengthscale
        else:
            self.covar_module = gpytorch.kernels.RBFKernel(
                    batch_shape=torch.Size([nb_gp]))

            if init_lengthscale_temporal is not None:

                if init_lengthscale_temporal == "mean":
                    if not same_inducing_points:
                        norme = torch.nn.functional.pdist(
                            inducing_points[0, :, :])
                    else:
                        norme = torch.nn.functional.pdist(
                            inducing_points[:, :])
                    mean = torch.mean(norme).item()
                    lengthscale = torch.as_tensor(mean)
                if init_lengthscale_temporal == "median":
                    if not same_inducing_points:
                        norme = torch.nn.functional.pdist(
                            inducing_points[0, :, :])
                    else:
                        norme = torch.nn.functional.pdist(
                            inducing_points[:, :])
                    median = torch.median(norme).item()
                    lengthscale = torch.as_tensor(median)
                if init_lengthscale_temporal == "dim":
                    dim = np.sqrt(inducing_points.size(-1))
                    lengthscale = torch.as_tensor(dim)
            self.covar_module.lengthscale = lengthscale

        self.SVDKL = SVDKL

    # Define a MultivariateNormal Distribution
    def forward(self, x):
        if self.SVDKL:
            x = x.transpose(-1, -2).unsqueeze(-1)
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        latent_pred = gpytorch.distributions.MultivariateNormal(mean_x,
                                                                covar_x)
        return latent_pred


def prediction_samples(model, likelihood, test_loader, samples):
    model.eval()
    likelihood.eval()

    observed_pred_list = []
    # Make predictions by feeding model through likelihood
    torch.random.manual_seed(2)
    with torch.no_grad(), gpytorch.settings.fast_pred_var(), gpytorch.settings.num_likelihood_samples(samples):

        for x_batch, y_batch in test_loader:
            observed_pred_inter = likelihood(model(x_batch))
            observed_pred = observed_pred_inter.probs.mean(0).argmax(-1)
            observed_pred = observed_pred.cpu()
            observed_pred_list.append(observed_pred)

    final_observed_pred = torch.cat(observed_pred_list)
    return final_observed_pred


def read_GP_json_file(json_model_file):
    if os.path.exists(json_model_file):
        # Opening JSON file
        with open(json_model_file, 'r') as openfile:
            print("OPEN JSON MODEL FILE ", json_model_file)
            # Reading from json file
            json_model = json.load(openfile)
            spatial = json_model['spatial']
            use_scaled_spatialdata = json_model['use_scaled_spatialdata']
            spatial_kernel = json_model['spatial_kernel']
            number_gp = json_model['number_gp']
            number_inducing_points = json_model['number_inducing_points']
            same_inducing_points = json_model['same_inducing_points']
            same_init_inducing_points = json_model['same_init_inducing_points']
            learning_inducing_points = json_model['learning_inducing_points']
            init_lengthscale_temporal = json_model['init_lengthscale_temporal']
            init_lengthscale_spatial = json_model['init_lengthscale_spatial']
            independent = json_model['independent']
            mixing_weights = json_model['mixing_weights']
            init_lmc = json_model['init_lmc']
            reduction = json_model['reduction']
            p = json_model['p']
            q = json_model['q']
            if "SVDKL" in json_model:
                SVDKL = json_model['SVDKL']
            else:
                SVDKL = False
            if "learning_outputscale" in json_model:
                learning_outputscale = json_model['learning_outputscale']
            else:
                learning_outputscale = True
            if "init_inducing_points" in json_model:
                init_inducing_points = json_model['init_inducing_points']
            else:
                init_inducing_points = False
            if "separate_loss" in json_model:
                separate_loss = json_model['separate_loss']
            else:
                separate_loss = False
            learning_rate = json_model['learning_rate']
            number_epochs = json_model['number_epochs']
            batch_size = json_model['batch_size']


            # Get values for each parameter
            name_model = json_model['name']
    else:
        exit("Model non defined, file {} doesn't exist".format(json_model_file))
    return name_model, spatial, use_scaled_spatialdata,\
           spatial_kernel, number_gp, number_inducing_points, same_inducing_points, \
           same_init_inducing_points, init_inducing_points, learning_inducing_points, \
           init_lengthscale_temporal, init_lengthscale_spatial, \
           learning_outputscale, independent, \
           mixing_weights, init_lmc, reduction, p, q, SVDKL, learning_rate, \
           number_epochs, batch_size, separate_loss

def main():

    json_model_file = "files/config_model.json"
    pt_data_file = "files/data.pt"
    pt_model_file = "files/model.pt"
    input_file = "files/stack.tif"

    new_samples = 10
    new_batch_size = 1024
    output_file = "classif_nb_batch_{}_nb_sample_{}.tif".format(new_batch_size, new_samples)

    if os.path.exists(input_file):

        name_model, spatial, use_scaled_spatialdata, \
        spatial_kernel, number_gp, number_inducing_points, same_inducing_points, \
        same_init_inducing_points, init_inducing_points, learning_inducing_points, \
        init_lengthscale_temporal, init_lengthscale_spatial, \
        learning_outputscale, independent, \
        mixing_weights, init_lmc, reduction, p, q, SVDKL, learning_rate, \
        number_epochs, batch_size, separate_loss = read_GP_json_file(json_model_file)

        checkpoint_data = torch.load(pt_data_file)

        sc_x_tem = checkpoint_data['sc_x_tem']
        le = checkpoint_data['encoder']
        nb_classes = checkpoint_data['nb_classes']

        with rasterio.open(input_file, 'r') as ds:
            arr = ds.read()
            profile = ds.profile
            profile.update(
                dtype=rasterio.uint8,
                count=1,
                compress='lzw')
        image = reshape_as_image(arr)
        X_tem = image.reshape(image.shape[0] * image.shape[1], image.shape[2])
        Y = np.ones(image.shape[0] * image.shape[1])
        X_scaled = sc_x_tem.transform(X_tem)

        if os.path.exists(pt_model_file):

            # Load training model
            checkpoint_GP = torch.load(pt_model_file,
                                       map_location=torch.device(
                                           'cpu'))
            inducing_points = checkpoint_GP['inducing_points']
            likelihoodSoftmax = checkpoint_GP['likelihood']

            nb_gp = nb_classes

            model_GP = GP_MultiClass(inducing_points,
                                     learning_inducing_points,
                                     nb_classes,
                                     nb_gp,
                                     same_inducing_points,
                                     independent,
                                     init_lengthscale_temporal,
                                     SVDKL,
                                     learning_outputscale)

            # Load GP model
            model_GP.load_state_dict(checkpoint_GP['model'])

            if not torch.is_tensor(X_scaled):
                X_scaled = torch.from_numpy(X_scaled).float()
            if not torch.is_tensor(Y):
                Y = torch.from_numpy(Y).long()

            if torch.cuda.is_available():
                X_scaled, Y = X_scaled.cuda(), Y.cuda()
                model_GP = model_GP.cuda()
                likelihoodSoftmax = likelihoodSoftmax.cuda()

            batch_size = new_batch_size
            dataset = torch.utils.data.TensorDataset(X_scaled, Y)
            test_loader = torch.utils.data.DataLoader(dataset,
                                                       batch_size=batch_size,
                                                       shuffle=False)

            # Prediction
            y_pred = prediction_samples(model_GP, likelihoodSoftmax,
                                        test_loader, new_samples)


            y_pred = le.inverse_transform(y_pred.numpy().astype(int))
            y_pred = y_pred.reshape(image.shape[0], image.shape[1])

            with rasterio.open(output_file, 'w', **profile) as dst:
                dst.write(y_pred.astype(rasterio.uint8), 1)
        else:
            print("Model not trained", pt_model_file)
            exit()

    else:
        print("File not found", input_file)


if __name__ == "__main__":
    main()

```

## Expected Behavior

The output is a `.tif` file. It can be open using the software QGIS.
 
The classification can vary depending on `new_number_samples` i.e. the number of samples to draw from a latent GP when computing a likelihood. However, the classification should not vary depending on `new_batch_size` i.e. the number of batchs.

All the files can be found here: 
[files.zip](https://github.com/cornellius-gp/gpytorch/files/11374347/files.zip)

Because of the length of the files, the file `stack.tif` is in comment.

## System information

**Please complete the following information:**
-  Gpytorch: 1.9.1
-  Torch: 1.12.1
-  Computer OS: CentOS Linux release 7.9.2009 (Core)

## Additional information

Because of the length of the files, the file `stack.tif` is in comment.

**Comment by valentinecesbio** [comment]
Stack file `stack.tif`: 
[file_stack.zip](https://github.com/cornellius-gp/gpytorch/files/11374387/file_stack.zip)

**Comment by gpleiss** [comment]
When making predictions with variational models, we draw from the joint latent distribution `p(f* | data)`. So the prediction dependent on the data in the mini batch. The result is smoother samples than what you would get if you drew from `p(f*_1 | data) ... p(f*_{Ntest} | data)` independently.

I could see an argument in favor of drawing from independent distributions for consistency, but the differences that you are seeing with different batch sizes are very slight, and well within what you would expect for a probabilistic program.
